File: dyna-cli/src/sync_client.rs
//! HTTP sync client for communicating with the remote Dyna server.
//!
//! This module encapsulates all network interactions between the CLI and the
//! remote server, providing a clean async API for changeset-based push, pull,
//! clone, promote, and channel management operations.
//!
//! All request bodies are gzip-compressed and sent with `Content-Encoding: gzip`.
//! The client also sends `Accept-Encoding: gzip` so the server can compress
//! responses. Responses are transparently decompressed (reqwest handles this
//! automatically when the feature is enabled, but we also handle it manually
//! for robustness).

use anyhow::{Context, Result};
use dyna_core::compression;
use dyna_core::protocol::*;
use reqwest::Client;

/// Client for interacting with the Dyna remote server.
pub struct SyncClient {
    client: Client,
    base_url: String,
}

impl SyncClient {
    /// Create a new sync client for the given remote URL.
    pub fn new(base_url: &str) -> Self {
        Self {
            client: Client::new(),
            base_url: base_url.trim_end_matches('/').to_string(),
        }
    }

    /// Generic helper: POST gzip-compressed JSON, check status, parse response.
    ///
    /// The request body is serialized to JSON, gzip-compressed, and sent with
    /// `Content-Encoding: gzip` and `Content-Type: application/json` headers.
    /// `Accept-Encoding: gzip` is also sent to request compressed responses.
    async fn post_json<Req, Resp>(
        &self,
        endpoint: &str,
        request: &Req,
        operation: &str,
    ) -> Result<Resp>
    where
        Req: serde::Serialize,
        Resp: serde::de::DeserializeOwned,
    {
        let url = format!("{}{}", self.base_url, endpoint);
        let compressed_body = compression::compress_json(request)
            .context(format!("Failed to compress {} request body", operation))?;

        let response = self
            .client
            .post(&url)
            .header("Content-Type", "application/json")
            .header("Content-Encoding", "gzip")
            .header("Accept-Encoding", "gzip")
            .body(compressed_body)
            .send()
            .await
            .context(format!(
                "Failed to connect to remote server for {}",
                operation
            ))?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("{} failed (HTTP {}): {}", operation, status, body);
        }

        // Read response bytes and transparently decompress
        let bytes = response
            .bytes()
            .await
            .context(format!("Failed to read {} response body", operation))?;

        let decompressed = compression::read_transparent(&bytes)
            .context(format!("Failed to decompress {} response", operation))?;

        serde_json::from_slice(&decompressed)
            .context(format!("Failed to parse {} response", operation))
    }

    /// Generic helper: GET with Accept-Encoding: gzip, check status, parse response.
    async fn get_json<Resp>(&self, endpoint: &str, operation: &str) -> Result<Resp>
    where
        Resp: serde::de::DeserializeOwned,
    {
        let url = format!("{}{}", self.base_url, endpoint);
        let response = self
            .client
            .get(&url)
            .header("Accept-Encoding", "gzip")
            .send()
            .await
            .context(format!(
                "Failed to connect to remote server for {}",
                operation
            ))?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("{} failed (HTTP {}): {}", operation, status, body);
        }

        let bytes = response
            .bytes()
            .await
            .context(format!("Failed to read {} response body", operation))?;

        let decompressed = compression::read_transparent(&bytes)
            .context(format!("Failed to decompress {} response", operation))?;

        serde_json::from_slice(&decompressed)
            .context(format!("Failed to parse {} response", operation))
    }

    /// Push changesets to the remote server.
    pub async fn push(&self, request: &PushRequest) -> Result<PushResponse> {
        self.post_json::<_, PushResponse>("/api/v1/push", request, "Push")
            .await
            .and_then(|resp| {
                resp.success
                    .then_some(resp.clone())
                    .ok_or_else(|| {
                        anyhow::anyhow!(
                            "Push rejected: {}",
                            resp.error.unwrap_or_else(|| "Unknown error".into())
                        )
                    })
            })
    }

    /// Pull changesets from the remote server.
    pub async fn pull(&self, request: &PullRequest) -> Result<PullResponse> {
        self.post_json("/api/v1/pull", request, "Pull").await
    }

    /// Clone a repository from the remote server.
    pub async fn clone_repo(&self, request: &CloneRequest) -> Result<CloneResponse> {
        self.post_json("/api/v1/clone", request, "Clone").await
    }

    /// List all channels on the remote server.
    pub async fn list_channels(&self) -> Result<ListChannelsResponse> {
        self.get_json("/api/v1/channels", "List channels").await
    }

    /// Create a new channel on the remote server.
    pub async fn create_channel(
        &self,
        request: &CreateChannelRequest,
    ) -> Result<CreateChannelResponse> {
        self.post_json("/api/v1/channels", request, "Create channel")
            .await
    }

    /// Promote changesets from one channel to another on the remote server.
    pub async fn promote(&self, request: &PromoteRequest) -> Result<PromoteResponse> {
        self.post_json::<_, PromoteResponse>("/api/v1/promote", request, "Promote")
            .await
            .and_then(|resp| {
                resp.success
                    .then_some(resp.clone())
                    .ok_or_else(|| {
                        anyhow::anyhow!(
                            "Promote rejected: {}",
                            resp.error.unwrap_or_else(|| "Unknown error".into())
                        )
                    })
            })
    }

    /// Check the health of the remote server.
    pub async fn health(&self) -> Result<HealthResponse> {
        self.get_json("/api/v1/health", "Health check").await
    }
}

File: dyna-go/dynago/sync_client.go
package dynago

import (
	"bytes"
	"compress/gzip"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

// SyncClient communicates with a dyna-server over HTTP.
type SyncClient struct {
	baseURL    string
	httpClient *http.Client
}

// NewSyncClient creates a SyncClient for the given server URL.
func NewSyncClient(baseURL string) *SyncClient {
	return &SyncClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// ---------------------------------------------------------------------------
// Push
// ---------------------------------------------------------------------------

// Push sends changesets to the remote server.
func (c *SyncClient) Push(ctx context.Context, req *PushRequest) (*PushResponse, error) {
	var resp PushResponse
	if err := c.postJSON(ctx, "/api/v1/push", req, &resp); err != nil {
		return nil, err
	}
	if !resp.Success {
		msg := "unknown error"
		if resp.Error != nil {
			msg = *resp.Error
		}
		return nil, fmt.Errorf("push rejected: %s", msg)
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Pull
// ---------------------------------------------------------------------------

// Pull fetches new changesets from the remote server.
func (c *SyncClient) Pull(ctx context.Context, req *PullRequest) (*PullResponse, error) {
	var resp PullResponse
	if err := c.postJSON(ctx, "/api/v1/pull", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Clone
// ---------------------------------------------------------------------------

// CloneRepo clones the entire repository from the remote server.
func (c *SyncClient) CloneRepo(ctx context.Context, req *CloneRequest) (*CloneResponse, error) {
	var resp CloneResponse
	if err := c.postJSON(ctx, "/api/v1/clone", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Channels
// ---------------------------------------------------------------------------

// ListRemoteChannels fetches all channels from the remote server.
func (c *SyncClient) ListRemoteChannels(ctx context.Context) (*ListChannelsResponse, error) {
	var resp ListChannelsResponse
	if err := c.getJSON(ctx, "/api/v1/channels", &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CreateRemoteChannel creates a channel on the remote server.
func (c *SyncClient) CreateRemoteChannel(ctx context.Context, req *CreateChannelRequest) (*CreateChannelResponse, error) {
	var resp CreateChannelResponse
	if err := c.postJSON(ctx, "/api/v1/channels", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Promote
// ---------------------------------------------------------------------------

// Promote promotes changesets from one channel to another on the remote.
func (c *SyncClient) Promote(ctx context.Context, req *PromoteRequest) (*PromoteResponse, error) {
	var resp PromoteResponse
	if err := c.postJSON(ctx, "/api/v1/promote", req, &resp); err != nil {
		return nil, err
	}
	if !resp.Success {
		msg := "unknown error"
		if resp.Error != nil {
			msg = *resp.Error
		}
		return nil, fmt.Errorf("promote rejected: %s", msg)
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Resource history
// ---------------------------------------------------------------------------

// ResourceHistory queries the change history of a specific resource.
func (c *SyncClient) ResourceHistory(ctx context.Context, resourceID string) (*ResourceHistoryResponse, error) {
	var resp ResourceHistoryResponse
	endpoint := fmt.Sprintf("/api/v1/resources/%s/history", resourceID)
	if err := c.getJSON(ctx, endpoint, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// Health
// ---------------------------------------------------------------------------

// Health checks the remote server health.
func (c *SyncClient) Health(ctx context.Context) (*HealthResponse, error) {
	var resp HealthResponse
	if err := c.getJSON(ctx, "/api/v1/health", &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ---------------------------------------------------------------------------
// WebSocket notifications
// ---------------------------------------------------------------------------

// NotificationHandler is called for each notification received.
type NotificationHandler func(data json.RawMessage)

// SubscribeNotifications connects to the server's WebSocket endpoint and
// calls handler for each notification. It blocks until the context is
// cancelled or the connection drops.
func (c *SyncClient) SubscribeNotifications(ctx context.Context, handler NotificationHandler) error {
	wsURL := strings.Replace(c.baseURL, "http://", "ws://", 1)
	wsURL = strings.Replace(wsURL, "https://", "wss://", 1)
	wsURL += "/api/v1/ws"

	conn, _, err := websocket.DefaultDialer.DialContext(ctx, wsURL, nil)
	if err != nil {
		return fmt.Errorf("websocket connect failed: %w", err)
	}
	defer conn.Close()

	done := make(chan struct{})
	go func() {
		defer close(done)
		for {
			_, message, err := conn.ReadMessage()
			if err != nil {
				return
			}
			handler(json.RawMessage(message))
		}
	}()

	select {
	case <-ctx.Done():
		conn.WriteMessage(websocket.CloseMessage,
			websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
		return ctx.Err()
	case <-done:
		return nil
	}
}

// ---------------------------------------------------------------------------
// HTTP helpers (with gzip compression)
// ---------------------------------------------------------------------------

func (c *SyncClient) postJSON(ctx context.Context, endpoint string, reqBody, respBody interface{}) error {
	data, err := json.Marshal(reqBody)
	if err != nil {
		return fmt.Errorf("marshal request: %w", err)
	}

	// Gzip compress
	var buf bytes.Buffer
	gz := gzip.NewWriter(&buf)
	if _, err := gz.Write(data); err != nil {
		return err
	}
	gz.Close()

	url := c.baseURL + endpoint
	req, err := http.NewRequestWithContext(ctx, "POST", url, &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Content-Encoding", "gzip")
	req.Header.Set("Accept-Encoding", "gzip")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("request to %s failed: %w", endpoint, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("%s failed (HTTP %d): %s", endpoint, resp.StatusCode, string(body))
	}

	body, err := readTransparent(resp)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}

	return json.Unmarshal(body, respBody)
}

func (c *SyncClient) getJSON(ctx context.Context, endpoint string, respBody interface{}) error {
	url := c.baseURL + endpoint
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept-Encoding", "gzip")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("request to %s failed: %w", endpoint, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("%s failed (HTTP %d): %s", endpoint, resp.StatusCode, string(body))
	}

	body, err := readTransparent(resp)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}

	return json.Unmarshal(body, respBody)
}

// readTransparent reads the response body, transparently decompressing gzip.
func readTransparent(resp *http.Response) ([]byte, error) {
	var reader io.Reader = resp.Body
	if resp.Header.Get("Content-Encoding") == "gzip" {
		gz, err := gzip.NewReader(resp.Body)
		if err != nil {
			return nil, err
		}
		defer gz.Close()
		reader = gz
	}
	return io.ReadAll(reader)
}
